### PR TITLE
Unit test

### PR DESCRIPTION
**Pull request by Zia123456**
# Unit-Test
- Unit-test created for following classes
- Book
- Person
- Student
- Teacher
- Classroom
- Rental
- all classes that was created while implementing the Decorator design pattern

### CODE REVIEW

File: spec/person_spec.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      expect(@person).to be_an_instance_of Person
+    end
+    it 'can_use_services when parent permission == true' do
+      expect(@person.can_use_services?).to eql true
+    end
+    it "Name should be 'Kennedy'" do
+      expect(@person.name).to eql 'Kennedy'
+    end
+  end
+end
```
Comment (bobb-Rob) [inline spec/person_spec.rb]: - [x] Based on the requirement, you are expected to test all the methods in all classes. In your Person_spec.rb file, i cannot see the test for `Person.add_rental` method. 

```ruby
def add_rental(date, book)
    Rental.new(date, book, self)
  end
```

- [x] Also, the method `can_use_services?` test should cover all scenario, 👇🏾 
    - Negative scenario - The `can_use_services?` method returns false if under age and no parent permission
    - Positive Scenario - The `can_use_services?` method returns true if over age
    - The can_use_services? method returns true if has parent permission - **_You have implemented this_**
Please implement that.

File: spec/classroom_spec.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+    it 'create a new Classroom' do
+      expect(classroom).to be_an_instance_of Classroom
+    end
+
+    it 'Should create a classroom' do
+      expect(classroom.label).to eq 'PostGate'
+    end
+  end
+end
```
Comment (bobb-Rob) [inline spec/classroom_spec.rb]: - [x] Similarly, I cannot see your test for `add_student` method in this classroom_spec.rb file. Please fix that.

File: spec/book_spec.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    end
+
+    it 'title off book should be Lord of the Ring' do
+      expect(book.title).to eq 'Lord of the Ring'
+    end
+
+    it 'Author of book should be Richard Benson' do
+      expect(book.author).to eq 'Richard Benson'
+    end
+  end
```
Comment (bobb-Rob) [inline spec/book_spec.rb]: - [x] Similarly, I do not see any test for `Book.add_rental` method in **_this_** test file. Please implement this.

File: spec/student_spec.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    it 'create a new student' do
+      expect(student).to be_an_instance_of Student
+    end
+
+    it 'the student can have permission to use the service' do
+      expect(student.can_use_services?).to eq true
+    end
+
+    it 'name of student should be James Brown' do
+      expect(student.name).to eq 'James Brown'
```
Comment (bobb-Rob) [inline spec/student_spec.rb:17]: - [x] I still do not see your test for `Student.add_classroom` method. Kindly add it.

```ruby
def add_classroom(classroom)
    @classroom = classroom
    classroom.student.push(self) unless classroom.student.include?(self)
  end
```

File: spec/teacher_spec.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+require_relative '../teacher'
+
```
Comment (bobb-Rob) [inline spec/teacher_spec.rb:3]: - [x] Please look through your **_classes_** again and ensure you implement tests for all the methods in them.

**Review by bobb-Rob — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️
Hi @Zia123456 and @dialite, 

Good job so far! 👏👏👏 
There are some issues that you still need to work on to go to the next project but you are almost there!  😃 
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


## Highlights
✅ No Linter error
✅ Correct use of Gitflow
✅ Very descriptive Pull Request(PR)
✅ Neat Commit messages

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/book_spec.rb (hunk 3 above)
Comment (dialite) [reply spec/book_spec.rb]: Thanks for the feedback. Book.add_rental method has been tested.

File: spec/person_spec.rb (hunk 1 above)
Comment (dialite) [reply spec/person_spec.rb]: Thanks for the feedback. All suggestions implemented

File: spec/classroom_spec.rb (hunk 2 above)
Comment (dialite) [reply spec/classroom_spec.rb]: Test for add_student have been added. Thanks

File: spec/teacher_spec.rb (hunk 5 above)
Comment (dialite) [reply spec/teacher_spec.rb:3]: All require test for all methods are done

File: spec/student_spec.rb (hunk 4 above)
Comment (dialite) [reply spec/student_spec.rb:17]: student.add_class method test is now included. Thank you

**Comment by dialite** [comment]
# keenly expecting to see this project approved!!!

**Review by franclobo — APPROVED** [review]
Hi @dialite, 

# STATUS: APPROVED 😃 

git-merge

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- All rspec tests passed ✔️ 
- All methods in classes are tested ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by dialite** [comment]
Thanks you for the positive feedback. We got what we expected.

**Pull request merged**